connectors/sources/servicenow.py

```diff
@@ -327,6 +434,111 @@ async def ping(self):
             logger.exception("Error while connecting to the ServiceNow.")
             raise
 
+    def _format_doc(self, data):
+        """Format document for handling empty values & type casting.
+
+        Args:
+            data (dict): Fetched record from ServiceNow.
+
+        Returns:
+            dict: Formatted document.
+        """
+
+        data = {key: value for key, value in data.items() if value}
+        data.update(
+            {
+                "_id": data["sys_id"],
+                "_timestamp": iso_utc(parser.parse(data["sys_updated_on"])),
+            }
+        )
+        return self.serialize(doc=data)
+
+    async def _attachment_fetcher(self, table_sys_id):
+        """Method to add attachment metadata to queue
+
+        Args:
+            table_sys_id (str): Table Id for fetching attachment metadata.
+        """
+
+        async for attachment_metadata in self.servicenow_client.fetch_data(
+            url=ENDPOINTS["ATTACHMENT"], params={"table_sys_id": table_sys_id}
+        ):
+            formatted_attachment_metadata = await asyncio.to_thread(
+                self._format_doc, data=attachment_metadata
+            )
+            await self.queue.put(
+                (  # pyright: ignore
+                    formatted_attachment_metadata,
+                    partial(
+                        self.servicenow_client.fetch_attachment_content,
+                        formatted_attachment_metadata,
+                    ),
+                )
+            )
+
+        await self.queue.put(END_SIGNAL)  # pyright: ignore
+
+    async def _producer(self, service_name):
+        """Fetch data for configured service name.
+
+        Args:
+            service_name (str): Service Name for preparing URL.
+        """
+
+        logger.debug(f"Fetching {service_name} data")
+        async for table_data in self.servicenow_client.fetch_data(
+            url=ENDPOINTS["TABLE"].format(table=service_name), params={}
+        ):
+            formatted_table_data = await asyncio.to_thread(
+                self._format_doc, data=table_data
+            )
+            await self.fetchers.put(
+                partial(
+                    self._attachment_fetcher,
+                    formatted_table_data["_id"],
+                )
+            )
+            self.task_count += 1
+
+            await self.queue.put((formatted_table_data, None))  # pyright: ignore
+        await self.queue.put(END_SIGNAL)  # pyright: ignore
+
+    async def _consumer(self):
+        """Consume the queue for the documents.
+
+        Yields:
+            dict: Formatted document.
+        """
+
+        while self.task_count > 0:
+            _, item = await self.queue.get()
+
+            if item == END_SIGNAL:
+                self.task_count -= 1
```

Comment:
question: For `_producer` we're incrementing the `task_count` for each iteration of the loop iterating over the table data, but we're only putting one `END_SIGNAL` into the queue. Will `task_count` be calculated correctly?

Reply:
Based on fetchers `task_count` are incremented & whenever that fetchers execution will be completed `END_SIGNAL` is added to the queue

Reply:
```
try:
      async for table_data in self.servicenow_client.get_table_data(
          url=ENDPOINTS["TABLE"].format(table=service_name)
      ):
          formatted_table_data = self._format_doc(data=table_data)
          serialized_table_data = self.serialize(doc=formatted_table_data)
          await self.fetchers.put(
              partial(
                  self._attachment_fetcher,
                  serialized_table_data["_id"],
              )
          )
          self.task_count += 1

          await self.queue.put((serialized_table_data, None))  # pyright: ignore
  except Exception as exception:
      logger.warning(
          f"Skipping table data for {service_name}. Exception: {exception}."
      )

  await self.queue.put(END_SIGNAL)  # pyright: ignore
```

I may misunderstand something, but if we look at this example for `_producer` I notice, that we increment the `task_count` for every iteration in the async loop, but we're only putting one `END_SIGNAL` into the queue.

So if `servicenow_client.get_table_data(...)` yields for example 5x `table_data` we've a `task_count` of 5 and one `END_SIGNAL`. The `_consumer` runs as long as `task_count` is greater than 0. As we're only having one `END_SIGNAL` in the queue the loop will run forever, right (we only decrement `task_count` once)? Am I missing something?

Reply:
`task_count` aren't updated based on async loop but based on `fetcher.put` we are incrementing.

Reply:
But `await self.queue.put(END_SIGNAL)` lives outside the loop in the example I've provided. So if `get_table_data` yields N times, we've N increments to `task_count`, but only one `END_SIGNAL`, right? Do we've an explicit test for this behavior?

Reply:
`END_SIGNAL` for every `task_count` are added to queue from the method which is going to be executed in `fetcher`

Reply:
Thanks for the explanation, we've agreed in a call to introduce an enumeration so we can differentiate between different `END_SIGNAL`(s). Then it's more obvious, which `END_SIGNAL` belongs to which task.